src/chainstate/stacks/boot/bns.clar

```diff
@@ -0,0 +1,919 @@
+    (is-eq char 0x6e) ;; n
+    (is-eq char 0x6f) ;; o
+    (is-eq char 0x70) ;; p
+    (is-eq char 0x71) ;; q
+    (is-eq char 0x72) ;; r
+    (is-eq char 0x73) ;; s
+    (is-eq char 0x74) ;; t
+    (is-eq char 0x75) ;; u
+    (is-eq char 0x76) ;; v
+    (is-eq char 0x77) ;; w
+    (is-eq char 0x78) ;; x
+    (is-eq char 0x79) ;; y
+    (is-eq char 0x7a))) ;; z
+
+(define-private (is-vowel (char (buff 1)))
+  (or 
+    (is-eq char 0x61) ;; a
+    (is-eq char 0x65) ;; e
+    (is-eq char 0x69) ;; i
+    (is-eq char 0x6f) ;; o
+    (is-eq char 0x75) ;; u
+    (is-eq char 0x79))) ;; y
+
+(define-private (is-special-char (char (buff 1)))
+  (or 
+    (is-eq char 0x2d) ;; -
+    (is-eq char 0x5f))) ;; _
+
+(define-private (is-char-valid (char (buff 1)))
+  (or 
+    (is-lowercase-alpha char)
+    (is-digit char)
+    (is-special-char char)))
+
+(define-private (is-nonalpha (char (buff 1)))
+  (or 
+    (is-digit char)
+    (is-special-char char)))
+
+(define-private (has-vowels-chars (name (buff 32)))
+  (> (len (filter is-vowel name)) u0))
+
+(define-private (has-nonalpha-chars (name (buff 32)))
+  (> (len (filter is-nonalpha name)) u0))
+
+(define-private (has-invalid-chars (name (buff 32)))
+  (< (len (filter is-char-valid name)) (len name)))
+
+(define-private (name-lease-started-at? (namespace-launched-at (optional uint)) 
+                                        (namespace-revealed-at uint)
+                                        (name-props (tuple 
+                                                  (registered-at (optional uint))
+                                                  (imported-at (optional uint))
+                                                  (revoked-at (optional uint))
+                                                  (zonefile-hash (buff 20)))))
+      (let ((registered-at (get registered-at name-props))
+            (imported-at (get imported-at name-props)))
+        ;; The namespace must be launched
+        (asserts! (is-some namespace-launched-at) (err ERR_NAMESPACE_NOT_LAUNCHED))
+        ;; Sanity check: the name must have been either be registered or imported
+        (asserts! (is-eq (xor 
+          (match (get registered-at name-props) res 1 0)
+          (match (get imported-at name-props)   res 1 0)) 1) (err ERR_PANIC))
+        ;; If the name was launched, then started-at will come from registered-at
+        (if (is-some registered-at)
+          ;; The name was registered - We return the registration block height
+          (ok (unwrap-panic registered-at))
+          (if (and (>= (unwrap-panic imported-at) namespace-revealed-at)
+                   (<= (unwrap-panic imported-at) (unwrap-panic namespace-launched-at)))
+            ;; The name was imported after revealing the namespace and before launching the namespace - We return the launch block height
+            (ok (unwrap-panic namespace-launched-at))
+            (err ERR_NAME_EXPIRED)))))
+
+(define-private (compute-name-price (name (buff 32))
```

Comment:
It might be a good idea to make this a `(define-read-only)` function, because there's a `/v1` API for getting the price that will need to call this method.

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/2013/commits/f3e2c482817f31c4e4500b5e8503f80e31a32da3